fiftyone/core/utils.py

```diff
@@ -1468,3 +1470,78 @@ def _load_until(self, goal_position):
                 current_position += self._bytes.write(next(self._iterator))
             except StopIteration:
                 break
+
+
+_SAFE_CHARS = set(string.ascii_letters) | set(string.digits) | set("-_.")
+_WHITESAPCE_CHARS = set(string.whitespace) | set("+")
+_URL_NAME_LENGTH_RANGE = (1, 100)
+
+
+def _sanitize_char(c):
+    if c in _SAFE_CHARS:
+        return c
+
+    if c in _WHITESAPCE_CHARS:
+        return "-"
+
+    return ""
+
+
+def to_url_name(name):
```

Comment:
I wish we could use a lib for this logic. it is pretty critical little piece of code and rather not maintain it :) https://www.npmjs.com/package/url-safe-string seems to possibly be able to cover most scenarios. totally up to you :)

Reply:
Hmm I assumed that there wouldn't be anything out there flexible enough to do what I had in mind, but it does look like that package does everything except stripping consecutive, leading, and trailing `-`.

Another point of discussion: I guess we'll want the URL validation logic always (only) in the backend, since it needs to run against data in the database to verify availability.

I added some more of my notes from last night when trying to decide what to propose for generating URL slugs for dataset names (I added it for saved views here, but we need to add it for dataset names too).

Any thoughts on whether you like the current logic (implementation aside)?

Reply:
I like it! `-` makes sense as a separator for url friendliness. curious why you chose to allow `.` and `_`. To me, it simpler and more readable to always enforce a single separator character, `-`.

Reply:
I definitely agree that a smaller character set for URLs is more readable.

The only consideration I had was that, if we were to convert `_.` to `-` and also convert everything to lowercase, for example (also more readable URLs) then a user would have a smaller set of available view names.

For example, if they name a view `my_view`, then they wouldn't be allowed to name another view `My View!`, because they'd both have URL name `my-view`.

Now, should the user really be using both `my_view` and `My View!`? We could definitely decide for them that they cannot do that, if folks feel strongly about it.